### ISSUE TITLE
Proposal: Tracer Components

### ISSUE THREAD

**Issue by discostu105** [bogdandrutu/openconsensus#223]
Proposal: Instead of a global tracer singleton, every library needs to create it's own tracer, which also has a proper name that identifies the library/component that uses the tracer.

Motivation:
 * Identify the library/component that created a span
 * Be able to turn the instrumentation of a certain library/component on/off. This might be necessary if the instrumentation
 ** causes too much overhead
 ** is buggy (e.g. does not call span.end in every case)
 ** avoid "double-tracing". E.g. a tracing vendor (e.g. Dynatrace) might already have built-in auto-instrumentation for a certain library. In this case the vendor might want to decide (automatically or config-based) which of the implementations should be enabled/disabled.

Technically, this could be achieved similar to how logging frameworks deal with Loggers and LoggerFactories ([log4j](https://www.slf4j.org/apidocs/org/slf4j/LoggerFactory.html)).

```
Tracer tracer = OpenConsensus.Tracing.getTracer("io.opensomething.contrib.mongo");
```

Not sure how a naming convention could look like. Not sure if "component" is the right word, as it has already been used in the past as attribute name. At Dynatrace we call that a "Sensor". Not sure about granularity of "Tracers". Should every class have one?

This is kind of related to the GlobalTracer discussion (#107).

**Comment by pavolloffay** [comment]
Some more thoughts, should every tracer in the registry support full configuration? For instance different sampling, reporting mechanism. Perf-wise I would prefer a cached registry and also a single instrumentation might use multiple instrumentation points - e.g. servlet filer and jaxrs interceptors. 

This could be an iteration after #107 - if we allow a globally scoped tracer.

**Comment by carlosalberto** [comment]
Something I'd be quite interested to know, is how common an application needs such specific scenario. I remembering seeing this pattern in HTrace, but haven't seen it much in other places. Depending on that, I'd be up to putting this in out `contrib/` artifact (as opposed to put it in the main API side).

**Comment by yurishkuro** [comment]
Love the use case. However, I'm not sure tracer-per-component is necessary to address that. For example, if instrumentation follows the OpenTracing conventions and sets the "component" tag on span builder, then the tracer can compare it with configuration and decide to return a noop span with span context equal to the parent.

**Comment by tedsuo** [comment]
We currently name components in OpenTracing via a tag:

| Span tag name | Type | Notes and examples |
|:--------------|:-----|:-------------------|
| `component` | string  | The software package, framework, library, or module that generated the associated Span. E.g., `"grpc"`, `"django"`, `"JDBI"`. |

I agree there should be a `component` scope, and it should be handled more directly – the tag approach was due to backwards compatibility, etc, and is not the best. 

This concept is more related to scope management and span creation; I don't feel it is related to the global tracer, or tracer instantiation. Tracers are a mechanism for recording data and exporting it - ideally, there should be only one. We should not conflate instantiating the export mechanism with name-spacing and scoping the current operations.

Likewise, I would prefer not to conflate this kind of namespacing to a particular language feature, such as classes or functions – it should be a cross-language concept.

This still leaves the question open as to where to put this information. I suggest that this information is part of the `Span` or `SpanContext`, and should be constructed there. Spans inherit their parents `component` unless it is overridden during construction.

**Comment by bogdandrutu** [comment]
Am I understanding correctly that we want to have a notion of `component` at the `SpanBuilder` time is ok, I think this achieves the end result of having `Spans` grouped by the `component`.

@tedsuo I don't think we should propagate this between spans, maybe require instrumentation to set this when they create new Spans.

**Comment by yurishkuro** [comment]
I can see pros and cons in inheriting component from parent span, but I think the con is bigger: if another layer doesn't set component then inheriting it is unexpected behavior. So I'd rather have it set on every span, which isn't too onerous for middleware for a given library/framework.

**Comment by bogdandrutu** [comment]
Also a contrib `TracerWithComponent implements Tracer` that simply set the component when a SpanBuilder is return is easy to offer.

**Comment by tedsuo** [comment]
@yurishkuro I'm not actually sure what the current inheritance behavior is with the `component` tag in OT.

@bogdandrutu that might be a good way to model the behavior @discostu105 was asking for. We could move it into core if it looks like a good pattern. Ideally, I would like "best practices" to not require contrib projects.

**Comment by carlosalberto** [comment]
I agree with @yurishkuro on *not* inheriting component from the parent `Span`. Doing so could get tricky eventually, and I'd stick with simplicity for now.

**Comment by tedsuo** [comment]
This is a bit broader, but relevant. Please bear with me:

In general, each trace event can be modeled as if it is occurring in a nested namespace. So far, I have seen a lot of interest in the following namespaces:

Trace::Service::Component::Operation

Currently state of affairs (correct me if I am wrong):
* The `trace` is modeled as a SpanContext, with an ID but no name or attributes.
* The `service` is modeled as a Tracer, with attributes but no name or ID.
* The `component` is proposed to be modeled as a Span property, with a name but no ID or attributes.
* The `operation` is modeled as a Span, with a name, ID, and attributes.

In the long run, I suspect we want something like Name, ID, and Attributes for each namespace.

In the short run, I am hearing a lot of interest in attributes for components. Only having them on the span and the service is proving cumbersome for real world usage.

So perhaps we should think deeper about this subject.

**Comment by austinlparker** [comment]
Would some sort of `ComponentScope` that can be passed into the span builder help here?

**Comment by yurishkuro** [comment]
I am not sure Name/ID/Attributes applies to all 4 entities:
  * `Name` for a trace is quite vague
  * `ID` for service/component is counterproductive (high cardinality)
  * Attributes make sense across all four entities (realizing that operation == span), however semantically different: static for service/component, dynamic for trace/span.

service and component sound like two different levels of `resource`

**Comment by tedsuo** [comment]
Thanks, I agree @yurishkuro, it's not clear that this needs to be fully normalized. But we have some holes, especially around attributes and the component scope.

**Comment by pavolloffay** [comment]
Instead of 

`Tracer tracer = OpenConsensus.Tracing.getTracer("io.opensomething.contrib.mongo");` we could simply have one global instance

`Tracer tracer = OpenConsensus.Tracing.getTracer();` hen tracer could be converted something like a builder which would modify some of the configurations - tags, exporter, sampling.

**Comment by jmacd** [comment]
I think developers want an API that lets them write concise instrumentation code, so that the weight of instrumentation doesn't overwhelm the code itself. The question is how individual modules of code will (concisely) declare their component in the source, at that point where spans are built. 

This suggests making it possible to build spans and supply very few "component" arguments, potentially just one argument at the call site. In my own code, I've accomplished this by bundling a set of attributes, including the component concept, into a single object that's passed to the span builder. Importantly, this includes more than one attribute potentially. 

The pattern generally involves some heavy-weight resource that is loaded into memory. It's part of a component, sure, but it also includes a number of other attributes that specifically describe an instance of the component being invoked at the point where a span is created. The code could read like:

```
class MyClass {
  TraceComponent traceComponent;
  
  MyClass() {
    traceComponent = OpenTelemetry.Tracing.getTracer().componentBuilder("this_is_my_class").
                                                withAttribute("key1", "value1").
                                                withAttribute("key2", "value2");
    // ...
  }

  void someMethod(String arg) {
     Span span = this.traceComponent.spanBuilder("someMethod").withAttribute("arg", arg).startSpan()
     // ...
  }
};
```

The component builder would be a subset of the span builder API, merely allowing attribute names and setting the standard component attribute, for use in building spans elsewhere in the same module of code.

**Comment by tedsuo** [comment]
Thanks @jmacd. I discussed this with @pavolloffay, @bogdandrutu, and @carlosalberto. 

An updated proposal is as follows:

```java
class MyLibrary {
  Tracer tracer;
  
  MyLibrary() {
    tracer = OpenTelemetry.getTracer().toBuilder("this_is_my_class").
                                                                setResource(resource).
                                                                build();
    // ...
  }

  void someMethod(String arg) {
     Span span = this.tracer.spanBuilder("someMethod").setAttribute("arg", arg).startSpan()
     // ...
  }
};
```

Main Notes:
* OpenTelemetry.getTracer() returns a `Tracer`. This `Tracer` is the base "Service Tracer" which was instantiated via a service loader, and contains the name and resources for the entire service.
* OpenTelemetry.getTracer().toBuilder() returns a `TracerBuilder` which allows a component (framework, mysql client, RPC library) to add a component name, and add additional resources which are associated only with that component, and not the rest of the service. This is considered a "Component Tracer." 

Additional Notes:
* We don't think we need an intermediary `OpenTelemetry.Tracing` class. The `OpenTelemetry` class can directly contain tracing, metrics, and tag methods.
* We'd like to remove the `setResource` method currently available on `Tracer`. Resources should only be set on tracer creation.

**Comment by tedsuo** [comment]
I don't know how the hell Github mis-ordered mine and @pavolloffay's comments, but it did.

**Comment by bogdandrutu** [comment]
@tedsuo see the attachment :))), @pavolloffay came from the future :)))
![Pavol_Is_In_The_Future](https://user-images.githubusercontent.com/1373887/57398220-56a7c200-7183-11e9-8904-7fa594641f9d.png)

**Comment by jmacd** [comment]
I like it, @tedsuo. I always focus on concision, so I wonder if a ResourceBuilder might help with readability. For example, to construct a resource in the same line of code, you might like to write:

```
tracer = OpenTelemetry.getTracer().toBuilder("this_is_my_class").
                                                                setResource(Resource.ofEntries(
                                                                       entry("key1", "value1"),
                                                                       entry("key2", "value2")
                                                               )).build();
```
